eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala

```diff
@@ -556,13 +556,21 @@ object Router {
    */
   case class PaymentContext(id: UUID, parentId: UUID, paymentHash: ByteVector32)
 
-  case class Route(amount: MilliSatoshi, hops: Seq[ChannelHop]) {
-    require(hops.nonEmpty, "route cannot be empty")
+  case class Route(amount: MilliSatoshi, hops: Seq[ChannelHop], finalHop_opt: Option[FinalHop]) {
+    require(hops.nonEmpty || finalHop_opt.nonEmpty, "route cannot be empty")
+
+    /** Full route including the final hop, if any. */
+    val fullRoute: Seq[Hop] = hops ++ finalHop_opt.toSeq
 
-    val length = hops.length
+    /**
+     * Fee paid for the trampoline hop, if any.
+     * Note that when using MPP to reach the trampoline node, the trampoline fee must be counted only once.
+     */
+    val trampolineFee: MilliSatoshi = finalHop_opt.collect { case hop: NodeHop => hop.fee(amount) }.getOrElse(0 msat)
 
-    def fee(includeLocalChannelCost: Boolean): MilliSatoshi = {
-      val hopsToPay = if (includeLocalChannelCost) hops else hops.drop(1)
+    /** Fee paid for the channel hops towards the recipient or the source of the final hop, if any. */
+    def channelFee(includeLocalChannelCost: Boolean): MilliSatoshi = {
+      val hopsToPay = if (includeLocalChannelCost) hops else hops.headOption.map(_ => hops.tail).getOrElse(Nil)
```

Comment:
No because `drop(1)` throws an exception if `hops` is `Nil` (which can happen since the requirement for `hops` to be `nonEmpty` is gone and replaced by `hops.nonEmpty || finalHop_opt.nonEmpty`.

Reply:
Fixed in https://github.com/ACINQ/eclair/pull/2480/commits/f7ec174629b3e19fce90a423ccf6ef93c2ffc86f, thanks!